### ISSUE TITLE
Seclude CLI params and service config resolution from the core

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#8364]
### Use case description

Ideally it should be straightforward to use Serverless Framework programmatically. Additionally clean separation of concerns will help significantly the maintenance and will allow to integrate Components engine into the repository.

_Improvements towards separation of Packaging and Deployment phases has been moved to https://github.com/serverless/serverless/issues/8499_

### Proposed solution

Step by step let's seclude CLI params and service config resolution logic from a core and layout a new processing flow as follows:

1. If CLI command is `-v`, output version info and abort
1. If in service context:
   1. Resolve service configuration (with support for `-c, --config` CLI params), into plain not normalized in any way JSON structure). (_If there's  parsing error and it's not a CLI help request, crash with meaningful error_)
   1. Resolve `file`, `self` and `strToBool` variable sources (but only those not depending on other resolvers)
   1. Ensure that `provider` and eventual `provider.stage` properties are fully resolved. (_If there's a validation error and it's not a CLI help request, crash with meaningful error on `provider` being not resolved. Show deprecation and warning notice on `provider.stage` not being resolved_)
   1. Load eventual extra env variables from `.env` files (_If there's  parsing error and it's not a CLI help request, crash with meaningful error_)
   1. Resolve `env` and remaining `file`, `self` and `strToBool` variable sources (but only those not depending on other resolvers)
   1. Ensure eventual `plugins` property is fully resolved.
   1. Initialize plugins (at this stage plugins should have no access to service config or any meta):
   (_If there's initialization error and it's not a CLI help request, crash with meaningful error_)
      1. Register plugin config schema extensions
      1. Register plugin variable extensions
      1. Register commands and hooks
1. If CLI help request. Display help and abort
1. Parse CLI arguments (with respect to map of supported commands and options)
1. If in service context:
   1. Resolve variables for all variable sources which do not depend on config properties 
   1. Resolve all remaining variables in service config
1. Run lifecycle events for CLI command

Additionally to have Framework CLI totally programatic, https://github.com/serverless/serverless/issues/1720 has to be addressed.

#### Implementation spec

Preliminary notes:
- Implementation of each step may break some tests. If those tests relate to functionalities being removed from core they should be simply removed. If it's otherwise ideally if they're fixed with refactoring to `runServerless` variant (assuming they're applicable for that) 
- Ideally if all below points are addressed after following tasks are addressed:
   - https://github.com/serverless/serverless/issues/8478 - Move all unit tests to `test/unit` folder
   - https://github.com/serverless/serverless/issues/8497 -  Remove CLI help `--verbose` option and improve general help output

##### 0.1 Unify process error handling

1. Wrap [all process logic](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/scripts/serverless.js#L16-L74) with try/catch clause
2. Configure handler which generally mirrors one at [`Error` class](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/lib/classes/Error.js#L67-L132) (but presents simplified logic due to more generic approach). Place handler logic in `lib/cli/handle-error.js` (and ensure some tests for it)
3. Ensure that [`uncaughtExceptions`](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/scripts/serverless.js#L24) are handled with same error handler
4. Remove [_lib/classes/Error.js_ `logError`](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/lib/classes/Error.js#L67-L132) utility

#####  0.2 Generalize process execution span promise handling (as currently served by `serverless.onExitPromise`)

1. Reflect execution span of introduced above `try/catch` clause in promise accessible at `lib/cli/execution-span.js` (module should export an unresolved _promise_ and method (to be used internally) for resolving it)
2. Assign `lib/cli/execution-span.js` promise to `serverless.executionSpan` (as it can be valuable for a plugins). Remove it's so far counterpart [`serverless.onExitPromise`](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/scripts/serverless.js#L38) and in places it was used, refer to either promise exported by `lib/cli/execution-span.js` or `serverless.executionSpan`
3. Move [`analytics.sendPending()`](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/scripts/serverless.js#L41-L43) to top of the try/catch clause

##### 1.0 Seclude `-v, --version` CLI params handling

1. At [begin of the flow](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/scripts/serverless.js#L15) implement following:
    - If CLI command is `sls -v [...]` or `sls --version [...]` Generate simple version output which follows one implemented internally in [`CLI` class](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/lib/classes/CLI.js#L359-L384). Place handling logic in `lib/cli/eventually-output-version-info.js` (and ensure some tests for it)
    - Abort the process (ensure no further process steps are pursued (but without hard exiting the process))
2. Remove `-v, --version` option handling and recognition from [CLI.js](https://github.com/serverless/serverless/blob/2fffb168bc7f957ed9e8e048fd08dfb9669e8eca/lib/classes/CLI.js) class

##### 2.0 Resolve eventual service config file path (service context)

1. Follow up with service context detection logic. It should be implemented in `lib/cli/resolve-service-config-path.js` and resemble logic we have now here: https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/utils/getServerlessConfigFile.js#L13-L48
2. Remove from internals any config file path resolution logic and ensure that above is handled as single point of truth:
   - Pass resolved config file path to [`Serverless` constructor](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/Serverless.js#L25) on `config.serviceConfigPath` and assign it to `serviceConfigPath` property
   - Resolve [`config.servicePath`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/Serverless.js#L42-L44) from `serviceConfigPath`
   - Remove [`findServicePath` util](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/Utils.js#L101-L123)
   - Replace `getConfigFilePath` usage with `resolveServiceConfigPath` at [interactive CLI setup](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/plugins/interactiveCli/initializeService.js#L8)
   - Replace any [`getServerlessConfigFilePath`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/utils/getServerlessConfigFile.js#L50-L55) usage with `serverless.serviceConfigPath` and remove this util entirely (in [interactive CLI setup](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/plugins/interactiveCli/initializeService.js#L83-L84) simply override `serverless.serviceConfigPath` instead)
    - Remove [`getConfigFilePath`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/utils/getServerlessConfigFile.js#L13-L48)

##### 2.1 Parse service config file source

1. Follow up with service config content resolution (bare resolution with no normalization or vars resolution at this point). It should be implemented in `lib/service-config/read-source.js`, and resemble logic we have here: https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/utils/getServerlessConfigFile.js#L112-L127. if `readServiceConfigSource` crashes expose the error only if it's not CLI help request, otherwise behave as we're not in service context.
2. Remove from internals any config source resolution:
   - Pass resolved config source to [`Serverless` constructor](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/Serverless.js#L25) on `config.serviceConfigSource` and assign it to `serviceConfigSource` property
   - In [`Servervless.js` class](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/Serverless.js) methods, replace `this.pluginManager.serverlessConfigFile` references with `this.serviceConfigSource`
   - In [interactive CLI setup](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/plugins/interactiveCli/initializeService.js#L83-L84) override `serverless.serviceConfigSource` with help of `readServiceConfigSource` module.
   - Remove  [`pluginManager.loadConfigFile` method](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L69-L83) and [`pluginManager.serverlessConfigFile`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L56) property

##### 2.2 Initial (partial) variables resolution

For that we would need to Implement new variables resolver with following modules:

###### `lib/variables/resolve-variables-map.js`

Function that takes `serviceConfig` as an input. Traverses it's properties and returns map of all properties which use variable syntax. Result map should expose all information needed for complete variables resolution without a need of repeated property parsing.

_After we will fully override variable resolution that's currently in a framework (point 5.2), function should be configured to also override all `serviceConfig` properties which depend on variables with `null` values (technically we _move_ all information to _variables map_, and remove it from `serviceConfig`. It's to ensure that eventual further `serviceConfig` processing in case of variable resolution errors is not affected by unresolved properties content)_

Expected format of result map

```javascript
const sep = "\0";

const exampleResultMap = {
  [`custom${sep}creds`]: {
    raw:
      '${file(../config.${opt:stage, self:provider.stage, "dev"}.json):CREDS}_${self:custom.foo}',
    meta: [
      // Start from last to first
      // If vars are part of a string, provide start and end locations
      // and unconditionally coerce result to string
      { start: 71, end: 89, sources: [{ source: 'self', address: { raw: 'custom.foo' } }] },
      {
        start: 0,
        end: 70,
        sources: [
          {
            source: 'file',
            param: {
              raw: '../config.${opt:stage, self:provider.stage, "dev"}.json',
              meta: [
                {
                  start: 10,
                  end: 50,
                  sources: [
                    { source: 'opt', address: { raw: 'stage' } },
                    { source: 'self', address: { raw: 'provider.stage' } },
                    { raw: 'dev' },
                  ],
                },
              ],
            },
            address: { raw: 'CREDS' },
          },
        ],
      },
    ],
  },
  [`layers${sep}hello${sep}path`]: {
    raw: '${self:custom.layerPath}',
    // If property value is entirely constructed with var
    // No start/end points need to be configured
    // In such case we also support any result type (no string coercion)
    variables: [{ sources: [{ source: 'self', address: { raw: 'custom.layerPath' } }] }],
  },
};
```

_Note: In case of resolution from external files, new added content will need to have eventual variables resolved through same util_

###### `lib/variables/resolve-variables.js`

Function that takes `serviceConfig`, `variablesMap` and `variablesResolvers` as an input.

`variablesResolvers` is expected to be a simple map with _source type_ as a key (e.g. `self, `file` etc.) and function that takes `serviceConfig` and eeventual _param_ configured for resolver as arguments. Function may return result _sync_ way or _async_ via returned _promise_

There should be attempt to resolve every property.
- If resolution succeed, resolved value should be assigned on `serviceConfig` object and _variable_ reference removed from `variablesMap`.
- if resolution for any would fail it should be reported by an error which trough code resembles why it failed (A. not supported source, B. invalid configuration, C. missing necessary input, D. external service error). Failed resolution attempt should be stored in `variablesMap` (in future processing, resolution should be reattempted only if fail was caused by _A_ error, in other cases there should be no retry.

If there's any fail. Function crashes, and on it's error it should expose `errorneusVariableKeys` property with keys to each variable resolutions that failed.

--- 

Having above:
1. Generate variables map
1. Attempt to resolve `file`, `self` and `strToBool` variable sources (but only those not depending on other resolvers). If it fails ignore any _not supported source_ errors. If there are other errors and it's not a CLI help request, in initial stage, ignore them, but after addressing 5.2 signal them with warning message and show a deprecation that with next major we will fail.

##### 2.3 Ensure `provider` and `provider.stage` properties are resolved.

1. Inspect variables map:
   - if `provider` property still depends on variable resolution, crash with meaningful error, that we cannot accept given form of configuration
   - if `provider.stage` property still depends on variable resolution. Show warning and deprecation, stating that it's not recommend to use variables at this property and that we will fail on that with next major

##### 2.4 Ensure to load env variables from `.env` files

Follow up with resolution of environment variables from `.env` files (currently being implemented at https://github.com/serverless/serverless/pull/8413)

##### 2.5 Further (partial) variables resolution

As in 2.1 step, attempt to resolve `file`, `self`,  `strToBool` and `env` variable sources (but only those not depending on other resolvers). If it fails ignore any _not supported source_ errors. If there are other errors and it's not a CLI help request in initial stage, ignore them, but after addressing 5.2 signal them with warning message and show a deprecation that with next major we will fail.

##### 2.6 Ensure eventual `plugins` property is fully resolved.

Inspect variables map, if `plugins` property still depends on variable resolution, crash with meaningful error, that we cannot accept given form of configuration

##### 2.7.0 Recognize help command

1. Implement _is help CLI command_ logic in `lib/cli/is-help-command.js` (it should follow [`cli.isHelpRequest` logic](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/CLI.js#L68-L84) but also recognize `--help-components`) and adapt it in internals:
    - Resolve `isHelpCommand` and pass it to `Serverless` constructor in `config.isHelpCommand` and internally assign it to `isHelpCommand` property
   - Replace internal `serverless.cli.isHelpRequest` usage with `serverless.isHelpCommand`
   - Replace internal [`pluginManager.cliOptions.help`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L149) usage with `serverless.isHelpCommand`
   - Remove [`serverless.cli.isHelpRequest` implementation](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/CLI.js#L68-L84)

##### 2.7.1 Recognize commands which are independent of external plugins

_Handling of those commands ideally should be totally secluded from Framework engine, still to not impose too timetaking refactor at this step let's simply mark them, to make further processing possible (having that addressed, let's open an issue calling for their seclusion)_

1. If CLI command is either `plugin`, `login`, `logout` or `dashboard` pass to `Serverless` constructor a `shouldMutePluginInitializationErrors: true` option, and internally assign t to  `_shouldMutePluginInitializationErrors` property
2. In [`pluginManager.resolveServicePlugins()`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L149) Rely on `serverles._shouldMutePluginInitializationErrors` and remove `pluginManager.pluginIndependentCommands` property.

##### 2.7.2 Initialize `Serverless` instance

_(this will most likely lay out naturally and should not require any code changes)_

Follow up with construction of `Serverless` instance and invoke of `serverless.init()`

##### 3.0 If CLI help command show help and abort

1. Implement display option help logic (to be used by various help variants) in `lib/cli/help/options.js`. It should take our common _command_ configuration object, and resemble logic we have at [`cli.displayCommandOptions()`](https://github.com/serverless/serverless/blob/786809e262b56490a78a923b0b031378badb18c0/lib/classes/CLI.js#L142-L182)
1. Implement interactive CLI help logic in `lib/cli/help/interactive.js`. It should be a function that accepts an _interactiveCLI_ command configuration and resembles logic we have at [`cli.generateInteactiveCliHelp()](https://github.com/serverless/serverless/blob/786809e262b56490a78a923b0b031378badb18c0/lib/classes/CLI.js#L258-L270)
1. Implement main CLI help logic in `lib/cli/help/framework.js`. It should be a function that accepts a `loadedPlugins` and resembles logic we have at [`cli.generateMainHelp()`](https://github.com/serverless/serverless/blob/786809e262b56490a78a923b0b031378badb18c0/lib/classes/CLI.js#L184-L256) (note we should have #8497 addressed at this point)
1. Implement specific command help logic in `lib/cli/help/command.js`. It should be a function that accepts `commandName` and `command` arguments, and:
   - [Dislpays plugin name](https://github.com/serverless/serverless/blob/786809e262b56490a78a923b0b031378badb18c0/lib/classes/CLI.js#L349-L350) if command is configured via external plugin
   - [Displays command usage and it's options](https://github.com/serverless/serverless/blob/786809e262b56490a78a923b0b031378badb18c0/lib/classes/CLI.js#L349-L350) - Note I believe we can implement inline [`cli.displayCommandUsage()`](https://github.com/serverless/serverless/blob/786809e262b56490a78a923b0b031378badb18c0/lib/classes/CLI.js#L123-L140) logic, and refer to already implemented `lib/cli/help/options.js`
1. if _interactive CLI_ help request. Find `InteractiveCli` plugin, run `lib/cli/help/interactive.js` with its comand and abort
1. If _general_ (not command specific) help request, run `lib/cli/help/framework.js` with `serverless.cli.loadedCommands`
1. If _command_ help request, find given command in `serverless.cli.loadedCommands`
   1. If specified command is not found show warning and output general help
   1. Otherwise run `lib/cli/help/command.js` with resolved comman
1. Remove following code:
   -  [`cli.displayHelp` call](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/Serverless.js#L147-L149) 
   - All help generation commands from [`CLI` class](https://github.com/serverless/serverless/blob/786809e262b56490a78a923b0b031378badb18c0/lib/classes/CLI.js)

##### 4.0 Parse CLI arguments

1. Having a map of all supported commands and options follow up with resolution of CLI arguments:
   - Parse CLI args with logic as in [`resolveCliInput`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/utils/resolveCliInput.js)
   - Ensure that all CLI args validation steps (aside of `validateServerlessConfigDependency` and `assignDefaultOptions`) as pursued in [`pluginManager.invoke`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L492-L497) are taken care of. Ideally if it's generalized, so can be also used to validate Components  CLI input
   -  Let's put it into `lib/cli/parse-params.js`
2. Pass resolved `commands` and `options` to `serverless.run()` method. In context `serverless.run()` assign those properties on`processedInput` property
   - To not break things for external plugins we cannot just remove resolution of `processedInput`, that happens [in `serverless.init()`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/Serverless.js#L73). Still let's override there `processedInput` with getter that exposes a deprecation message if property is accessed at initialization phase (having that we will remove it next major)
   - In each internal plugin remove handling of second constructor option (CLI options) - (it'll also automatically address other important concern -> https://github.com/serverless/serverless/issues/2582). If for some reason reliance on CLI options seems crucial at implementation phase, then move its handling to `initialize` lifecycle hook (it's first lifecycle event propagated unconditionally). Access CLI options from `serverless.processedInput` (and treat it as read only)
   - Refactor [`pluginManager.validateOptions`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L572-L599) so it's eventual errors do not refer to CLI params (this method will now be effective only for programmatic usage)
   - Refactor [`pluginManager.validateServerlessConfigDependency`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L560-L570) so it's eventual errors do not refer to CLI usage (e.g. we should refer to _service context_ and not to _service directory_)
   - Remove [`pluginManger.convertShortcutsIntoOptions`](https://github.com/serverless/serverless/blob/bde334c10f33a813a16e158832eb61c56e77d822/lib/classes/PluginManager.js#L632-L644) as Framework will already be populated with resolved shortcuts

##### 5.1 Resolve variables for all variable sources which do not depend on config properties

As in 2.1 step, attempt to resolve all variable sources which do not depend on config properties.

If it fails ignore any _not supported source_ errors. If there are other errors in initial stage, ignore them, but after addressing 5.2 signal them with warning message and show a deprecation that with next major we will fail.

##### 5.2 Resolve all remaining variables in service config

As in 2.1 step, attempt to resolve all remaining variables.

If it fails signal them with warning message and show a deprecation that with next major we will fail. Additionally:
- Ensure that after resolving variables map (point 2.2) all service config properties configure through variables are preset to `null`
- In any variable resolution step, convert errors ignoring to warnings with and deprecation that with next major we will fail.

Remove all variable resolution logic from Framework core

##### 6.0 Run lifecycle events for CLI command

_(this will most likely lay out naturally and should not require any code changes)_

Follow up with `serverless.run()`

--- 

### _Progress summary:_
- [x] _0.0.1 - https://github.com/serverless/serverless/issues/8478 - Move all unit tests to `test/unit` folder_
- [x] _0.0.2 - https://github.com/serverless/serverless/issues/8497 - Remove CLI help `--verbose` option and improve general help output_
- [x] _0.1.0 - Unify process error handling_
- [x] _0.2.0 - Generalize process execution span promise handling_
- [x] _1.0.0 - Seclude `-v`, `--version` CLI params handling_
- [x] _2.0.0 - Resolve eventual service config file path (service context)_
- [x] _2.1.0 - Parse service config file source_
- [x] _2.2.0 - Initial (partial) variables resolution_
- [x] _2.3.0 - Ensure provider and provider.stage properties are resolved._
- [x] _2.4.0 - Ensure to load env variables from .env files_
- [x] _2.5.0 - Further (partial) variables resolution_
- [x] _2.6.0 - Ensure eventual plugins property is fully resolved_
- [x] _2.7.0 - Recognize help command_
- [x] _2.7.1 - Recognize commands which are independent of external plugins_
- [x] _2.7.2 - Initialize Serverless instance_
- [x] _3.0.0 - If CLI help command show help and abort_
- [x] _4.0.0 - Parse CLI arguments_
- [x] _5.1.0 - Resolve variables for all variable sources which do not depend on config properties_
- [ ] _5.2.0 - Resolve all remaining variables in service config_
- [x] _6.0.0 - Run lifecycle events for CLI command_

**Comment by pgrzesik** [comment]
Hello @medikoo :wave: I've started thinking about this refactoring, starting with the very first step: `0.1 Unify process error handling` and I'd like to ensure that my thinking is correct and to clarify a few things. 

1. By `Wrap all process logic with try/catch clause`, do you have in mind wrapping the whole highlighted logic with the following approach? ```
(async () => {
  try {
    ...
  } catch {
  ...
  }
})()
```
2. What do you exactly mean by simplified logic/generic approach to that handler? What should be simplified - that is not clear to me (yet)?
3. How do you imagine still supporting `enterpriseErrorHandler` in that solution? By nesting try/catch clauses?

Thanks in advance :bow:

**Comment by medikoo** [comment]
@pgrzesik great thanks for asking. This issue was not specified in deep detail, as we planned to handled that internally (intentionally I didn't put "help wanted" label at ths point), as some parts are challenging, and it needs to be done carefuly  with other aspects in mind (like future components adaptation)

I've also actually also started work on my side on that (but didn't place PR yet) :)

Anyway let me answer your questions:

> By Wrap all process logic with try/catch clause, do you have in mind wrapping the whole highlighted logic with the following approach?

It's about covering most of Framework CLI logic with async try/catch clause (yes, as you show in your snippet) and handle any error with unified handler.

> What do you exactly mean by simplified logic/generic approach to that handler? 

I think I envisioned a simpler output, but as I was implementing this, I didn't go any further than refactoring this logic so it works as standalone. I guess in this initial phase, there's not much else to simplify (probably later when will want to cover also Components with same handler, some output will have to be updated)

> How do you imagine still supporting enterpriseErrorHandler in that solution?

At this point I would leave it working same way. I would just not mute original error if handler crashes on it's own (I don't think that current logic is right)

**Comment by pgrzesik** [comment]
Thanks for a comprehensive explanation @medikoo and sorry - I've missed the fact that this ticket didn't have "help wanted" label :bow:

**Comment by neverendingqs** [comment]
I was doing some investigation on how the changes here will impact [`serverless-dotenv-plugin`](https://github.com/neverendingqs/serverless-dotenv-plugin). I did some quick tests with https://github.com/serverless/serverless/pull/8987 and it looks like `env` variables are currently resolved before and after step 2.7 at `serverless>=2.26.0`.

If I am understanding the proposed solution correctly, `env` variables will eventually only get resolved at step 2.5. If that is the case, is it safe to say that any `env` variables loaded by `serverless-dotenv-plugin` will eventually not get resolved because they're not loaded early enough in the process?

**Comment by medikoo** [comment]
@neverendingqs In new variables resolver `env` variables are only resolved from environment variables as found on `process.env`. Still if configuration will reference variables not defined in `process.env`, then properties will be kept unresolved, and it's old variables resolver that will attempt to resolve them (and it's where I believe `serverless-dotenv-plugin` will jump in with it's resolution).

So if `serverless-dotenv-plugin` does not override eventual system environment variables with other values, then nothing should change in context of v2 of Framework.

However with v3 of a Framework old resolver will be removed, and then Framework will crash on unresolved env variables. I think right solution for that would be to configure a new variables source (e.g `lambdaEnv`) for new variables resolver in `serverless-dotenv-plugin`, and then in configuration use `lambdaEnv` instead of `env` where intention is to resolve for lambda configuration environment variables with help of `serverless-dotenv-plugin`

**Comment by medikoo** [comment]
@neverendingqs note that we're open to offer any help needed to adapt the plugin to new resolver. So do not hesitate to ask, we may also discuss it further in context of `serverless-dotenv-plugin` repository (cc @pgrzesik )

**Comment by neverendingqs** [comment]
Thanks @medikoo. I think for v3, it might be best to try to propose / port the popular use cases of the plugin into Serverless's dotenv resolver, and then deprecate the plugin, as I don't think auto-loading all envs for all Lambda functions is safe from a security point of view.

**Comment by neverendingqs** [comment]
If auto-loading env vars for all functions is desirable, I can always adjust the plugin to loop through all the functions and add it, so I don't believe the plugin will need the proposed `lambdaEnv`.

**Comment by neverendingqs** [comment]
Ah also, if `file` loads first before `env`, then technically users can do whatever they want by loading `dotenv` in a JavaScript config file as long as they import at least one variable from that file. That could be an immediate workaround for anyone in v3.

**Comment by medikoo** [comment]
> if `file` loads first before `env`

Technically it's more that all properties are resolved at once, and if property which is behind variable depends on the other behind variable, then the other is ensured to be fully resolved before we attempt to resolve first one.

**Comment by neverendingqs** [comment]
> @neverendingqs note that we're open to offer any help needed to adapt the plugin to new resolver.

Hi @medikoo (cc @pgrzesik). If you have a chance, I would appreciate your feedback on doing something like this to support more complicated `dotenv` workflows in Serverless, to make sure I am understanding the new variables engine correctly:

* `serverless.yml`: https://github.com/neverendingqs/serverless-dotenv-plugin/pull/153/files#diff-36b3f719786f0fd2a95cbd33132afd135906fb79fecd7c985cf8c7f0ecb01eb0
* `configs.js`: https://github.com/neverendingqs/serverless-dotenv-plugin/pull/153/files#diff-d808b034dd73fa8224e04137fc8d16beb0dc37ee672fcb38fcc85e86b0aa42c8

I think this is the best way forward into `serverless>=3.0.0`, instead of continuing support of `serverless-dotenv-plugin`.

**Comment by neverendingqs** [comment]
I'm not sure if this should go in its own issue, but thoughts on `.env` files (2 iv) being resolved before `file` (2 ii)? That would allow environment variables defined in `.env` be used in files (e.g. 

**configs.js**
```js
module.exports = async ({ options, resolveConfigurationProperty }) => {
  const region = options.region
    || process.env.REGION
    || 'us-east-1;
  ...
}
```

**.env**
```
REGION=us-east-2
```
).

**Comment by medikoo** [comment]
@neverendingqs I believe we fixed issue you were confused by, at https://github.com/serverless/serverless/pull/9110

In general there's no order in which variables are resolved. All are resolved at once, and if depedency is detected, the dependent waits for dependency.

In case were you were approaching error, problem was that at initial resolution phase we assume env variables not being fulfilled (due to eventual `.env` not being loaded yet), and it's the reason JS function resolver crashed.
Bug on our side was that we exposed that crash. Proper behavior is to silently catch it, and re-run resolver after `.env` is loaded. https://github.com/serverless/serverless/pull/9110 puts that on track.

**Comment by neverendingqs** [comment]
Thanks! I understand what is happening better now, but am still unsure how the example I provided above will resolve with the changes in #9110.

If there are no `${env:REGION}` references in `serverless.yml`, will `region` resolve to `us-east-1` or `us-east-2`? I suspect it can resolve to `us-east-1` today because `process.env.REGION` won't cause an error.

Having it call `dotenv` to load all environment variables in `.env` first is handy because we can then read and/or manipulate environment variables in JavaScript files, either for making decisions on overrides like in the example or to parse the value first (e.g. convert a CSV string into an array).


**.env**
```
REGION=us-east-2
```

**configs.js**
```js
module.exports = async ({ options, resolveConfigurationProperty }) => {
  const region = options.region  // Not provided in `options`
    || process.env.REGION        // Defined in `.env`
    || 'us-east-1';              // Default if not found in either `options` or `.env`

  // Convert an env var CSV string into an array before returning
  const statements = (process.env.ALLOWED_PRINCIPALS || []).map(...);
  ...
}
```

**Comment by medikoo** [comment]
> If there are no ${env:REGION} references in serverless.yml, will region resolve to us-east-1 or us-east-2?

AWS Region in a framework is resolved as: 
- If `--region` option was passed with a CLI command use it
- If `provider.region` was set in service configuration use it
- Use `us-east-1`

No environment variable influences that resolution.

Still you've pointed one aspect that's currently missed when JS function resolvers are in place.

Indeed when such resolver is invoked first time, `.env` files are not loaded yet, and currently there are no clear means paved on how to abort such invocation, so it's re-triggered after `.env` is loaded (technically it can be hacked by throwing an error with `MISSING_VARIABLE_DEPENDENCY` code, but that's dirty).

I've proposed on how to solve it at https://github.com/serverless/serverless/issues/9112 and it's not challenging to do. We will implement that shortly, as definitely it's needed to consider JS resolvers functionality complete.

**Comment by medikoo** [comment]
@neverendingqs experience with JS function resolvers should be improved with next version, as we've just merged this: https://github.com/serverless/serverless/pull/9140

**Comment by neverendingqs** [comment]
Thanks!

it looks like `dotenv` is loaded after variable resolution is attempted:

* https://github.com/serverless/serverless/blob/ee01833df80db8bc751f8248f49944738606393f/scripts/serverless.js#L151-L166
* https://github.com/serverless/serverless/blob/ee01833df80db8bc751f8248f49944738606393f/scripts/serverless.js#L232-L233

Is this because we need to resolve `useDotenv` first before we can determine if we should use dotenv?

Could we avoid a resolution of `env` variables if we instead loaded environment variables from dotenv first? As a side effect, it would also resolve the use case I described in https://github.com/serverless/serverless/issues/8364#issuecomment-798829324 and https://github.com/serverless/serverless/issues/8364#issuecomment-799499482.

**Comment by medikoo** [comment]
> Is this because we need to resolve useDotenv first before we can determine if we should use dotenv?

Yes, and we need to resolve _stage_ (which depends on `provider.stage`) to be able to determine eventual `.env.<stage>` filename.

> Could we avoid a resolution of env variables if we instead loaded environment variables from dotenv first? 

Why? What exactly problem will it solve? Note that `.env` won't, override already set environment variables, so it's safe to resolve those already found in environment.

**Comment by neverendingqs** [comment]
> Why? What exactly problem will it solve? Note that `.env` won't, override already set environment variables, so it's safe to resolve those already found in environment.

Using the below as an example:

```js
module.exports = async ({ options, resolveConfigurationProperty }) => {
  // Convert an env var CSV string into an array of e.g. IAM statements before returning
  const statements = (process.env.ALLOWED_PRINCIPALS || [])
    .split(',')
    .map(principal => ({ ... } );
  return { statements };
}
```

If `ALLOWED_PRINCIPALS` is never set in the system and only defined in `dotenv` we will always return

```js
{ statements: [] }
```

today. This means we can't use different `.env.{stage}` files to set different `ALLOWED_PRINCIPALS` values.

Note: this is just a use case I would be interested in, but totally understandable if it's not one that will be supported.

**Comment by medikoo** [comment]
@neverendingqs with next release JS function resolver will be invoked before `.env` files are loaded __only__ if it plays part in resolution of either `provider.stage` or `useDotenv` properties. Otherwise it'll invoked after `.env` files are loaded. Just that I believe will fix you're use case

And in case JS function resolver plays part of resolution of `provider.stage` or `useDotenv` then obviously it needs to be called before `.env` files are loaded, but then it also should not internally rely on env vars, as that creates an obvious conflict.
